### PR TITLE
Packaging v2

### PR DESCRIPTION
**Pull request by simutisernestas**
hey, i've updated the packaging format as you suggested. Evaluation and training scripts are left in the root directory while the package itlsef is moved to the yolact_edge directory. This way we can keep all functionality of those nice scripts and have portable package for folks to use it in scripts. Also these changes are on the laters master branch. I guess the old branch of packaging can go and we could create new one for this PR or merge it to master if you're happy with the result. Thanks!

### CODE REVIEW

**Comment by haotian-liu** [comment]
Thank you so much on your work in packaging, I will be looking into this soon.

**Comment by SteveMacenski** [comment]
@haotian-liu I also took a look over this - I'm happy with these changes and I think they should be merged :-)

**Review by haotian-liu — CHANGES_REQUESTED** [review]
Thank you so much for your implementation. They look great! There are a few concerns I have which I left as comments (you could share your thoughts over the comments). Furthermore, I believe `./docker/`, `data/scripts/`, `data/*.gif` should be moved to the root directory.

One major concern I have is about the `config.py`. I am still unsure about how we should deal with this (as this might be a internal issue of the original design). Ideally, the specific configs should be in the root directory (say under `./data/config.py`), while how `cfg` are initialized should be in the `package_root/data/config.py`. Maybe one way of doing is to implement a copy function, where we have a default cfg in `package_root/data/config.py` served as the global. And we can replace that with any config objects we later have. What's your opinion towards this?

Some other files in `./scripts` might also need to be moved but for which file speficially in `scripts` folder should be moved, we can discuss later.

Thank you so much again for contributing to our code base.

**Comment by haotian-liu** [comment]
@SteveMacenski Thank you for also reviewing this. The contribution looks good, while I have some concerns over a part of it. I left the comments in the review above, and it will be really appreciated if you would like to share your thoughts too. Thanks.

File: test.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from yolact_edge.layers.output_utils import postprocess, undo_image_transformation
+import pycocotools
+from yolact_edge.yolact import Yolact
+
+
+
+
+if __name__ == '__main__':
+  y = Yolact()
+
```
[the remark this answers is not in the document]
Comment (simutisernestas) [reply test.py:17]: no idea how i've missed it 😄

File: yolact_edge/layers/functions/detection.py [hunk 2]
```diff
@@ -257,7 +257,7 @@ def traditional_nms(self, boxes, masks, scores, iou_threshold=0.5, conf_thresh=0
             if cls_scores.size(0) == 0:
                 continue
             
-            preds = torch.cat([boxes[conf_mask], cls_scores[:, None]], dim=1).cpu().numpy()
+            preds = torch.cat([boxes[conf_mask], cls_scores[:, None]], dim=1).detach().cpu().numpy()
```
[the remark this answers is not in the document]
Comment (simutisernestas) [reply yolact_edge/layers/functions/detection.py:260]: when doing inference Pytorch is complaining about it and says that you should detach the vector before moving it to CPU (can't recall the actual error, but it's something along those lines). Do you know how to solve that in another way?

File: .gitignore [hunk 3]
```diff
@@ -156,4 +156,5 @@ data/MOTS
 data/FlyingChairs
 data/sbd
 data/video_files
-data/image_files
\ No newline at end of file
+data/image_files
+calib_images
```
[the remark this answers is not in the document]
Comment (simutisernestas) [reply .gitignore]: done

File: yolact_edge/layers/functions/detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (simutisernestas) [reply yolact_edge/layers/functions/detection.py]: I've removed this one at the moment. Would be nice to hear your comment on this issue anyways. I'm still wondering why I encountered this error.

File: README.md [hunk 4]
```diff
@@ -146,6 +146,62 @@ Use the help option to see a description of all available command line arguments
...
+datasets = [
+    'coco2014_dataset',
+    'coco2017_dataset',
+    'coco2017_testdev_dataset',
+    'flying_chairs_dataset',
+    'youtube_vis_dataset',
+]
+dataset = datasets[1]
+# Used tensorrt calibration
+calib_images = "/home/ernestas/test/calib_images"
```
[the remark this answers is not in the document]
Comment (simutisernestas) [reply README.md]: done

File: README.md [hunk 5]
```diff
@@ -146,6 +146,62 @@ Use the help option to see a description of all available command line arguments
...
 
+### Programmatic inference
+
+There are two steps to make this work:
+ 1) Install YOLACT edge as python package: ```pip install .```
+ 2) Use it in your own script: 
+
+```Python
+# from yolact_edge.inference import YOLACTEdgeInference
+import time
```
[the remark this answers is not in the document]
Comment (simutisernestas) [reply README.md]: i've created ```pkg_usage.py```, couldn't think of a better name 😄

File: README.md [hunk 6]
```diff
@@ -146,6 +146,62 @@ Use the help option to see a description of all available command line arguments
...
+    'yolact_edge_vid_trainflow_config',
+    'yolact_edge_youtubevis_config',
+    'yolact_resnet50_config',
+    'yolact_resnet152_config',
+    'yolact_edge_resnet50_config',
+    'yolact_edge_vid_resnet50_config',
+    'yolact_edge_vid_trainflow_resnet50_config',
+    'yolact_edge_youtubevis_resnet50_config',
+]
+config = model_configs[-5]
```
[the remark this answers is not in the document]
Comment (simutisernestas) [reply README.md]: i was counting from the other end of the list 😅  fixed this too

File: test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (simutisernestas) [reply test.py]: deleted

File: yolact_edge/layers/functions/detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (haotian-liu) [reply yolact_edge/layers/functions/detection.py]: It was most probably due to that you did not use `with torch.no_grad()` and the returned tensors are some part of the computational graph (with gradients)?

**Comment by haotian-liu** [comment]
Not sure if you have finished updating the PR, but there were some comments that I could not add in the files. Thanks!

> Furthermore, I believe ./docker/, data/scripts/, data/*.gif should be moved to the root directory.
> 
> One major concern I have is about the config.py. I am still unsure about how we should deal with this (as this might be a internal issue of the original design). Ideally, the specific configs should be in the root directory (say under ./data/config.py), while how cfg are initialized should be in the package_root/data/config.py. Maybe one way of doing is to implement a copy function, where we have a default cfg in package_root/data/config.py served as the global. And we can replace that with any config objects we later have. What's your opinion towards this?
> 
> Some other files in ./scripts might also need to be moved but for which file speficially in scripts folder should be moved, we can discuss later.

File: yolact_edge/layers/functions/detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (simutisernestas) [reply yolact_edge/layers/functions/detection.py]: indeed, that's the case. Thanks!

**Comment by simutisernestas** [comment]
Yeah, i've missed those. I'll get back to the comment soon.

**Comment by simutisernestas** [comment]
The original implementation of config.py is quite nasty, any time i've touched it while playing with code everything starts breaking. As I understand under the current situation everthing would work fine and the configuration contained in that file only specifies static parameters for different weights and datasets. What would be the benefits of moving it to root directory? 

IMHO ideal case would be to have a proper config file (for instance `YAML` format) which would be required as input alongside the model weights and expose only dynamic parameters which should be modified by the user. However, it would require quite some effort i guess.

I still have to take a second look at this later to make reasonable assessment.

**Review by haotian-liu — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Awesome! This will make releasing binaries of this _much_ easier in the ROS ecosystem so we can leverage it in the navigation system dynamic obstacle integrations! Once we have some demos / things merged in, we'll follow up to show you how your work is making a real impact in mobile robotics world wide once the integration is complete!

**Comment by haotian-liu** [comment]
Merged. Thank you so much in supporting our work!
@simutisernestas @SteveMacenski

**Comment by SteveMacenski** [comment]
Thank @simutisernestas, I'm just an observer / motivator :1st_place_medal:
